File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/AgentPipeline.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.Collections.Generic
open System.Text
open System.Text.Json
open Azure.AI.OpenAI

type BearerTokenPolicy(token: string) =
    inherit Azure.Core.Pipeline.HttpPipelineSynchronousPolicy()
    override _.OnSendingRequest(message: Azure.Core.HttpMessage) =
        message.Request.Headers.SetValue("Authorization", sprintf "Bearer %s" token)

module AgentPipeline =

    type PartialToolCall = {
        Id: ToolCallId option
        Name: ToolName option
        ArgsBuilder: StringBuilder
    }

    let private emptyPartialToolCall = {
        Id = None
        Name = None
        ArgsBuilder = StringBuilder()
    }

    let private upsertPartialToolCall (index: int) (idOpt: ToolCallId option) (nameOpt: ToolName option) (argsFragment: string) (partials: Map<int, PartialToolCall>) =
        let current = partials |> Map.tryFind index |> Option.defaultValue emptyPartialToolCall
        let nextId = if idOpt.IsSome then idOpt else current.Id
        let nextName = if nameOpt.IsSome then nameOpt else current.Name
        if not (String.IsNullOrEmpty(argsFragment)) then
            current.ArgsBuilder.Append(argsFragment) |> ignore

        partials
        |> Map.add index {
            Id = nextId
            Name = nextName
            ArgsBuilder = current.ArgsBuilder
        }

    let private toToolCalls (partials: Map<int, PartialToolCall>) : ToolCall list =
        partials
        |> Seq.sortBy (fun kv -> kv.Key)
        |> Seq.choose (fun kv ->
            match kv.Value.Id, kv.Value.Name with
            | Some id, Some name ->
                Some {
                    Id = id
                    Name = name
                    ArgumentsJson = kv.Value.ArgsBuilder.ToString()
                }
            | _ -> None)
        |> Seq.toList

    let aggregateStreamResponse (stream: System.Collections.Generic.IAsyncEnumerable<StreamChunk>) : Async<Result<LlmTurnResponse, StreamAggregationError>> =
        async {
            let textBuilder = StringBuilder()
            let mutable hasCompleted = false
            let mutable finishReason = ""
            let mutable partialToolCalls = Map.empty<int, PartialToolCall>

            try
                let enumerator = stream.GetAsyncEnumerator()
                let mutable keepReading = true

                while keepReading do
                    let! hasNext = enumerator.MoveNextAsync().AsTask() |> Async.AwaitTask
                    if not hasNext then
                        keepReading <- false
                    else
                        match enumerator.Current with
                        | TextDelta content ->
                            if not (String.IsNullOrEmpty(content)) then
                                textBuilder.Append(content) |> ignore
                        | ToolCallDelta (index, idOpt, nameOpt, argsFragment) ->
                            partialToolCalls <- upsertPartialToolCall index idOpt nameOpt argsFragment partialToolCalls
                        | StreamCompleted reason ->
                            hasCompleted <- true
                            finishReason <- reason

                let text = textBuilder.ToString()
                let toolCalls = toToolCalls partialToolCalls

                if hasCompleted then
                    let normalized = if isNull finishReason then "" else finishReason.Trim().ToLowerInvariant()
                    if normalized = "length" then
                        return Error (PartialResponse text)
                    else
                        return Ok { Content = text; ToolCalls = toolCalls }
                else
                    if String.IsNullOrWhiteSpace(text) then
                        return Error (PartialResponse "")
                    else
                        return Error (PartialResponse text)
            with _ ->
                let partialText = textBuilder.ToString()
                return Error (PartialResponse partialText)
        }

    let streamToLlmResponse (streamingCaller: StreamingLlmCaller) (schemas: ToolSchema list) (msgs: AgentMessage list) : Async<Result<LlmTurnResponse, LlmError>> =
        async {
            let! streamResult = streamingCaller schemas msgs
            match streamResult with
            | Error err -> return Error err
            | Ok stream ->
                let! aggregated = aggregateStreamResponse stream
                match aggregated with
                | Ok response -> return Ok response
                | Error (PartialResponse partialText) when not (String.IsNullOrWhiteSpace(partialText)) ->
                    return Ok { Content = partialText; ToolCalls = [] }
                | Error _ ->
                    return Error NoChoicesReturned
        }

    /// Step 2.1a: Pre-API Check - Interrupt Guard
    let checkInterrupt (state: TurnState) : StepResult<TurnState, TurnResult> =
        match state.Command with
        | InterruptTurn ->
            printfn "  [Turn Exit] Turn interrupted by user."
            Exit {
                Outcome = TurnOutcome.Interrupted
                Messages = state.Messages
                ApiCalls = state.ApiCalls
            }
        | RunTurn ->
            Continue state

    /// Step 2.1b: Pre-API Check - Budget Guard
    let checkBudget (state: TurnState) : StepResult<TurnState, TurnResult> =
        if state.ApiCalls >= state.Config.MaxIterations then
            printfn "  [Turn Exit] Reached max iterations (%d)." state.Config.MaxIterations
            Exit {
                Outcome = TurnOutcome.Failed (FailureReason.BudgetExhausted "Budget exhausted")
                Messages = state.Messages
                ApiCalls = state.ApiCalls
            }
        else
            Continue { state with ApiCalls = state.ApiCalls + 1 }

    /// Step 2.2: Prepare Messages Payload (shallow copy pipeline step)
    let prepareApiMessages (msgs: AgentMessage list) : AgentMessage list =
        msgs |> List.map id

    /// Step 2.3: Context Window Protection & Compaction Engine (Pure Pipeline Transformation)
    let compressContextIfNeeded (limit: int) (msgs: AgentMessage list) : AgentMessage list =
        ContextCompressor.compress 0.80 limit msgs

    /// Pipeline composition for message payload preparation
    let preparePayload (limit: int) (msgs: AgentMessage list) : AgentMessage list =
        msgs
        |> prepareApiMessages
        |> compressContextIfNeeded limit

    let llmErrorMessage (err: LlmError) =
        match err with
        | NoChoicesReturned -> "No choices returned from LLM"
        | ApiCallFailed message -> message

    /// Step 2.4: Execute API Call with Exponential Backoff Retry (Pure Async Recursion)
    let rec callLlmWithRetry (llmCaller: LlmCaller) (schemas: ToolSchema list) (maxRetries: int) (retryCount: int) (msgs: AgentMessage list) : Async<Result<LlmTurnResponse, LlmError>> =
        async {
            let! result = llmCaller schemas msgs
            match result with
            | Ok response -> return Ok response
            | Error err ->
                printfn "  [API Error Retry %d/%d] %s" (retryCount + 1) maxRetries (llmErrorMessage err)
                if retryCount >= maxRetries - 1 then
                    return Error err
                else
                    let delayMs = (int (Math.Pow(2.0, float retryCount))) * 1000
                    do! Async.Sleep delayMs
                    return! callLlmWithRetry llmCaller schemas maxRetries (retryCount + 1) msgs
        }

    let private cleanJsonArgs (argsStr: string) =
        let mutable clean = if String.IsNullOrWhiteSpace argsStr then "{}" else argsStr.Trim()

        let extractBraces () =
            let braceIndex = clean.IndexOf('{')
            if braceIndex > 0 then
                let lastBrace = clean.LastIndexOf('}')
                if lastBrace > braceIndex then
                    clean <- clean.Substring(braceIndex, lastBrace - braceIndex + 1)

        extractBraces ()

        if clean.StartsWith("\"") && clean.EndsWith("\"") then
            try
                let unescaped = JsonSerializer.Deserialize<string>(clean)
                if not (String.IsNullOrWhiteSpace unescaped) then clean <- unescaped.Trim()
            with _ -> ()

        extractBraces ()

        if not (clean.StartsWith("{")) then
            let rawString = clean.Trim('"')
            let dict = dict [ "path", rawString; "command", rawString; "key", rawString; "task", rawString; "text", rawString; "url", rawString ]
            clean <- JsonSerializer.Serialize dict

        clean

    /// Step 2.6: Process Tool Calls (Self-Correction, JSON Validation & Execution)
    let processToolCalls (executor: ToolExecutor) (registeredNamesSet: Set<ToolName>) (content: string) (toolCalls: ToolCall list) (state: TurnState) : Async<TurnState> =
        async {
            let! newToolMessages =
                toolCalls
                |> Seq.map (fun toolCall ->
                    async {
                        let name = toolCall.Name
                        let callId = toolCall.Id
                        let argsStr = cleanJsonArgs toolCall.ArgumentsJson
                        
                        let nameStr = ToolName.value name

                        if not (registeredNamesSet.Contains(name)) then
                            let avail = registeredNamesSet |> Seq.map ToolName.value |> String.concat ", "
                            let errStr = sprintf "Error: Tool '%s' is not registered. Available tools: %s" nameStr avail
                            printfn "  [Tool Validation Error] %s" errStr
                            return ToolMessage(callId, errStr)
                        else
                            try
                                use doc = JsonDocument.Parse(argsStr)
                                printfn "  [Tool Execution] %s(%s)" nameStr argsStr
                                let swTool = System.Diagnostics.Stopwatch.StartNew()
                                let! execResult = executor name argsStr
                                swTool.Stop()
                                AgentTelemetry.trackToolExecution state.SessionId state.UserId state.TurnIndex nameStr swTool.ElapsedMilliseconds argsStr execResult (Some state.SessionId) (Some state.TurnSpanId)
                                printfn "  [Tool Result] %s" execResult
                                return ToolMessage(callId, execResult)
                            with jsonEx ->
                                let errStr = sprintf "Error: Invalid JSON arguments for tool '%s': %s" nameStr jsonEx.Message
                                printfn "  [JSON Parse Error] %s" errStr
                                return ToolMessage(callId, errStr)
                    })
                |> Async.Parallel

            let mutationNames = ["file_write"; "write_to_file"; "file_patch"; "replace_file_content"]
            let verificationNames = ["read_terminal"; "terminal_execute"; "run_tests"; "dotnet_test"]

            let executedNames = toolCalls |> List.map (fun tc -> ToolName.value tc.Name)
            let hasMutation = state.HasFileMutations || (executedNames |> List.exists (fun n -> List.contains n mutationNames))
            let hasVerification = 
                if executedNames |> List.exists (fun n -> List.contains n verificationNames) then true
                elif executedNames |> List.exists (fun n -> List.contains n mutationNames) then false
                else state.HasExecutedVerification

            let assistantMsg = AssistantMessage(content, toolCalls)
            let updatedHistory = state.Messages @ (assistantMsg :: (newToolMessages |> Array.toList))
            return { state with Messages = updatedHistory; HasFileMutations = hasMutation; HasExecutedVerification = hasVerification }
        }

    /// Step 2.7: Process Final Text Response with Empty Response Recovery & Pre-Verify Gate
    let processTextResponse (rawContent: string) (state: TurnState) : StepResult<TurnState, TurnResult> =
        let finalText = if isNull rawContent then "" else rawContent.Trim()

        if String.IsNullOrEmpty finalText then
            if state.EmptyContentRetries < 2 then
                printfn "  [Empty Response Recovery] Retrying with prompt nudge..."
                let updatedHistory = state.Messages @ [ UserMessage "Please provide a complete text response summarizing your answer." ]
                Continue { state with Messages = updatedHistory; EmptyContentRetries = state.EmptyContentRetries + 1 }
            else
                let fallbackText = "(empty response)"
                printfn "Assistant: %s" fallbackText
                let updatedHistory = state.Messages @ [ AssistantMessage(fallbackText, []) ]
                Exit {
                    Outcome = TurnOutcome.Completed fallbackText
                    Messages = updatedHistory
                    ApiCalls = state.ApiCalls
                }
        else
            if state.HasFileMutations && not state.HasExecutedVerification && state.PreVerifyNudges < 2 then
                printfn "  [Pre-Verify Quality Gate] Intercepted completed turn. Files modified without verification. Prompting agent to verify..."
                let nudgeMsg = UserMessage "You modified files during this turn. Please verify your changes by executing unit tests or build commands before concluding."
                let updatedHistory = state.Messages @ [ AssistantMessage(finalText, []); nudgeMsg ]
                Continue { state with Messages = updatedHistory; PreVerifyNudges = state.PreVerifyNudges + 1 }
            else
                AgentTelemetry.trackTurnEnd state.SessionId state.UserId state.TurnIndex 0L finalText "completed" (Some state.SessionId) (Some state.TurnSpanId)
                let updatedHistory = state.Messages @ [ AssistantMessage(finalText, []) ]
                Exit {
                    Outcome = TurnOutcome.Completed finalText
                    Messages = updatedHistory
                    ApiCalls = state.ApiCalls
                }

    /// Step 2.1b: Pre-API Steering Drain (/steer)
    let drainSteering (queue: System.Collections.Concurrent.ConcurrentQueue<string>) (history: AgentMessage list) : AgentMessage list =
        if isNull queue then history else
        let items = System.Collections.Generic.List<string>()
        let mutable item = ""
        while queue.TryDequeue(&item) do
            if not (System.String.IsNullOrWhiteSpace item) then
                items.Add(item)

        if items.Count = 0 then
            history
        else
            let steeringContent = "\n\n[USER STEERING INTERRUPT]: " + String.concat "\n" items
            printfn "  [Pre-API Steering Drain] Injected %d mid-turn steering message(s)" items.Count

            match List.rev history with
            | ToolMessage(callId, content) :: rest ->
                let updatedToolMsg = ToolMessage(callId, content + steeringContent)
                List.rev (updatedToolMsg :: rest)
            | UserMessage content :: rest ->
                let updatedUserMsg = UserMessage(content + steeringContent)
                List.rev (updatedUserMsg :: rest)
            | _ ->
                history @ [ UserMessage ("[USER STEERING INTERRUPT]: " + String.concat "\n" items) ]

    /// Pure, Tail-Recursive 4-Phase Turn Loop Function
    let rec runTurnLoop (llmCaller: LlmCaller) (executor: ToolExecutor) (registeredSchemas: ToolSchema list) (registeredNamesSet: Set<ToolName>) (state: TurnState) : Async<TurnResult> =
        async {
            // Pipeline Step 2.1: Pre-API Checks (Interrupt Guard -> Budget Guard)
            match checkInterrupt state with
            | Exit result -> return result
            | Continue stateAfterInterruptCheck ->

            match checkBudget stateAfterInterruptCheck with
            | Exit result -> return result
            | Continue stateAfterBudgetCheck ->

            // Step 2.1b: Pre-API Steering Drain
            let stateDrained =
                if not (isNull stateAfterBudgetCheck.SteeringQueue) then
                    let updatedMessages = drainSteering stateAfterBudgetCheck.SteeringQueue stateAfterBudgetCheck.Messages
                    { stateAfterBudgetCheck with Messages = updatedMessages }
                else stateAfterBudgetCheck

            // Pipeline Step 2.2 & 2.3: Message Payload Preparation & Context Compression
            let preparedPayload = preparePayload stateDrained.Config.ContextWindowLimit stateDrained.Messages

            // Pipeline Step 2.4: Inner LLM API Call with Retry
            let swLlm = System.Diagnostics.Stopwatch.StartNew()
            let! apiResult = callLlmWithRetry llmCaller registeredSchemas stateDrained.Config.MaxRetries 0 preparedPayload
            swLlm.Stop()

            match apiResult with
            | Error NoChoicesReturned ->
                let message = llmErrorMessage NoChoicesReturned
                let res = { Outcome = TurnOutcome.Failed (FailureReason.NoResponse message); Messages = stateDrained.Messages; ApiCalls = stateDrained.ApiCalls }
                AgentTelemetry.trackTurnEnd stateDrained.SessionId stateDrained.UserId stateDrained.TurnIndex 0L message "no_response" (Some stateDrained.SessionId) (Some stateDrained.TurnSpanId)
                return res
            | Error (ApiCallFailed err) ->
                let res = { Outcome = TurnOutcome.Failed (FailureReason.ApiError err); Messages = stateDrained.Messages; ApiCalls = stateDrained.ApiCalls }
                AgentTelemetry.trackTurnEnd stateDrained.SessionId stateDrained.UserId stateDrained.TurnIndex 0L err "api_error" (Some stateDrained.SessionId) (Some stateDrained.TurnSpanId)
                return res
            | Ok response ->
                AgentTelemetry.trackLlmCall stateDrained.SessionId stateDrained.UserId stateDrained.TurnIndex "gpt-4o" swLlm.ElapsedMilliseconds response.Content response.ToolCalls (Some stateDrained.SessionId) (Some stateDrained.TurnSpanId)

                // Pipeline Step 2.6: Tool Call Execution Path
                if response.ToolCalls.Length > 0 then
                    let! nextState = processToolCalls executor registeredNamesSet response.Content response.ToolCalls stateDrained
                    return! runTurnLoop llmCaller executor registeredSchemas registeredNamesSet nextState

                // Pipeline Step 2.7: Final Text Response Path
                else
                    match processTextResponse response.Content stateDrained with
                    | Exit turnResult ->
                        let respText = match turnResult.Outcome with TurnOutcome.Completed txt -> txt | TurnOutcome.Failed (FailureReason.ApiError e) -> e | _ -> ""
                        AgentTelemetry.trackTurnEnd stateDrained.SessionId stateDrained.UserId stateDrained.TurnIndex 0L respText "completed" (Some stateDrained.SessionId) (Some stateDrained.TurnSpanId)
                        return turnResult
                    | Continue nextState -> return! runTurnLoop llmCaller executor registeredSchemas registeredNamesSet nextState
        }

File: implementations/csharp/src/Skight.AgentPlatform/AgentPipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Azure.AI.OpenAI;

namespace Skight.AgentPlatform
{
    public class TurnResult
    {
        public string FinalResponse { get; set; } = string.Empty;
        public List<ChatRequestMessage> Messages { get; set; } = new();
        public int ApiCalls { get; set; }
        public bool Completed { get; set; }
        public bool Failed { get; set; }
        public bool Interrupted { get; set; }
        public string ExitReason { get; set; } = "unknown";
        public string? Error { get; set; }
    }

    public class AgentPipeline
    {
        private readonly OpenAIClient _client;
        private readonly string _model;
        private readonly ToolRegistry _registry;

        public AgentPipeline(OpenAIClient client, string model, ToolRegistry registry)
        {
            _client = client;
            _model = model;
            _registry = registry;
        }

        private List<ChatRequestMessage> PrepareApiMessages(List<ChatRequestMessage> msgs)
        {
            return new List<ChatRequestMessage>(msgs);
        }

        private List<ChatRequestMessage> CompressContextIfNeeded(List<ChatRequestMessage> msgs, int contextWindowLimit)
        {
            return ContextCompressor.Compress(0.80, contextWindowLimit, msgs);
        }

        public async Task<TurnResult> RunTurnLoopAsync(AgentSessionState session, AgentConfig config, Func<List<FunctionDefinition>, List<ChatRequestMessage>, Task<ChatCompletions>>? customLlmCaller = null)
        {
            int apiCalls = 0;
            int emptyContentRetries = 0;
            string turnSpanId = Guid.NewGuid().ToString("N");
            var promptText = (session.Messages.LastOrDefault() as ChatRequestUserMessage)?.Content ?? string.Empty;
            AgentTelemetry.TrackTurnStart(session.SessionId, session.UserId, session.TurnCount, promptText, session.SessionId, turnSpanId);
            var swTurn = System.Diagnostics.Stopwatch.StartNew();

            while (apiCalls < config.MaxIterations)
            {
                if (session.InterruptRequested)
                {
                    session.InterruptRequested = false;
                    Console.WriteLine("  [Turn Exit] Turn interrupted by user.");
                    swTurn.Stop();
                    var turnRes = new TurnResult { FinalResponse = string.Empty, Messages = session.Messages, ApiCalls = apiCalls, Interrupted = true, ExitReason = "interrupted" };
                    AgentTelemetry.TrackTurnEnd(session.SessionId, session.UserId, session.TurnCount, swTurn.ElapsedMilliseconds, turnRes.FinalResponse, turnRes.ExitReason, session.SessionId, turnSpanId);
                    return turnRes;
                }

                apiCalls++;

                DrainSteering(session);

                var preparedMessages = CompressContextIfNeeded(PrepareApiMessages(session.Messages), config.ContextWindowLimit);

                ChatCompletions? completions = null;
                Exception? lastError = null;
                var swLlm = System.Diagnostics.Stopwatch.StartNew();

                for (int retry = 0; retry < config.MaxRetries; retry++)
                {
                    try
                    {
                        var toolDefinitions = _registry.GetToolSchemas();
                        if (customLlmCaller != null)
                        {
                            completions = await customLlmCaller(toolDefinitions.ToList(), preparedMessages);
                        }
                        else
                        {
                            var options = new ChatCompletionsOptions(_model, preparedMessages) { Temperature = 0.7f };

                            foreach (var toolDef in toolDefinitions)
                            {
                                options.Tools.Add(new ChatCompletionsFunctionToolDefinition(toolDef));
                            }

                            var response = await _client.GetChatCompletionsAsync(options);
                            completions = response.Value;
                        }
                        break;
                    }
                    catch (Exception ex)
                    {
                        lastError = ex;
                        Console.WriteLine($"  [API Error Retry {retry + 1}/{config.MaxRetries}] {ex.Message}");
                        if (retry == config.MaxRetries - 1)
                        {
                            swTurn.Stop();
                            var errRes = new TurnResult { FinalResponse = string.Empty, Messages = session.Messages, ApiCalls = apiCalls, Failed = true, ExitReason = "api_error", Error = ex.Message };
                            AgentTelemetry.TrackTurnEnd(session.SessionId, session.UserId, session.TurnCount, swTurn.ElapsedMilliseconds, errRes.FinalResponse, errRes.ExitReason, session.SessionId, turnSpanId);
                            return errRes;
                        }
                        await Task.Delay((int)Math.Pow(2, retry) * 1000);
                    }
                }

                swLlm.Stop();

                if (completions == null || completions.Choices.Count == 0)
                {
                    swTurn.Stop();
                    var noChoiceRes = new TurnResult { FinalResponse = string.Empty, Messages = session.Messages, ApiCalls = apiCalls, Failed = true, ExitReason = "no_response", Error = lastError?.Message ?? "No choices returned." };
                    AgentTelemetry.TrackTurnEnd(session.SessionId, session.UserId, session.TurnCount, swTurn.ElapsedMilliseconds, noChoiceRes.FinalResponse, noChoiceRes.ExitReason, session.SessionId, turnSpanId);
                    return noChoiceRes;
                }

                var choice = completions.Choices[0];
                var message = choice.Message;

                AgentTelemetry.TrackLlmCall(session.SessionId, session.UserId, session.TurnCount, _model, swLlm.ElapsedMilliseconds, message.Content ?? string.Empty, message.ToolCalls, session.SessionId, turnSpanId);

                if (message.ToolCalls != null && message.ToolCalls.Count > 0)
                {
                    var assistantMessage = new ChatRequestAssistantMessage(message.Content);
                    var registeredSchemas = _registry.GetToolSchemas();
                    var registeredNames = new HashSet<string>(registeredSchemas.Select(s => s.Name));

                    foreach (var toolCall in message.ToolCalls)
                    {
                        if (toolCall is ChatCompletionsFunctionToolCall functionCall)
                        {
                            assistantMessage.ToolCalls.Add(functionCall);
                            var name = functionCall.Name;
                            var callId = functionCall.Id;

                            if (!registeredNames.Contains(name))
                            {
                                var errorMsg = $"Error: Tool '{name}' is not registered. Available tools: {string.Join(", ", registeredNames)}";
                                Console.WriteLine($"  [Tool Validation Error] {errorMsg}");
                                AgentTelemetry.TrackToolExecution(session.SessionId, session.UserId, session.TurnCount, name, 0, functionCall.Arguments, errorMsg, session.SessionId, turnSpanId, isError: true);
                                session.Messages.Add(new ChatRequestToolMessage(errorMsg, callId));
                                continue;
                            }

                            string cleanArgs = SanitizeJsonArguments(name, functionCall.Arguments);

                            try { using var doc = JsonDocument.Parse(cleanArgs); }
                            catch (Exception jsonEx)
                            {
                                var errorMsg = $"Error: Invalid JSON arguments for tool '{name}': {jsonEx.Message}";
                                Console.WriteLine($"  [JSON Parse Error] {errorMsg}");
                                AgentTelemetry.TrackToolExecution(session.SessionId, session.UserId, session.TurnCount, name, 0, functionCall.Arguments, errorMsg, session.SessionId, turnSpanId, isError: true, exception: jsonEx);
                                session.Messages.Add(new ChatRequestToolMessage(errorMsg, callId));
                                continue;
                            }

                            Console.WriteLine($"  [Tool Execution] {name}({cleanArgs})");

                            try
                            {
                                var mutationTools = new HashSet<string> { "file_write", "write_to_file", "file_patch", "replace_file_content" };
                                var verificationTools = new HashSet<string> { "read_terminal", "terminal_execute", "run_tests", "dotnet_test" };

                                if (mutationTools.Contains(name))
                                {
                                    session.HasFileMutations = true;
                                    session.HasExecutedVerification = false;
                                }
                                else if (verificationTools.Contains(name))
                                {
                                    session.HasExecutedVerification = true;
                                }

                                var swTool = System.Diagnostics.Stopwatch.StartNew();
                                var result = await _registry.ExecuteToolAsync(name, cleanArgs);
                                swTool.Stop();
                                AgentTelemetry.TrackToolExecution(session.SessionId, session.UserId, session.TurnCount, name, swTool.ElapsedMilliseconds, cleanArgs, result, session.SessionId, turnSpanId);
                                Console.WriteLine($"  [Tool Result] {result}");
                                session.Messages.Add(new ChatRequestToolMessage(result, callId));
                            }
                            catch (Exception execEx)
                            {
                                var errorResult = $"Error executing tool '{name}': {execEx.Message}";
                                Console.WriteLine($"  [Tool Runtime Error] {errorResult}");
                                AgentTelemetry.TrackToolExecution(session.SessionId, session.UserId, session.TurnCount, name, 0, cleanArgs, errorResult, session.SessionId, turnSpanId, isError: true, exception: execEx);
                                session.Messages.Add(new ChatRequestToolMessage(errorResult, callId));
                            }
                        }
                    }
                    session.Messages.Insert(session.Messages.Count - message.ToolCalls.Count, assistantMessage);
                    continue;
                }

                var finalText = (message.Content ?? string.Empty).Trim();

                if (string.IsNullOrEmpty(finalText))
                {
                    if (emptyContentRetries < 2)
                    {
                        emptyContentRetries++;
                        Console.WriteLine("  [Empty Response Recovery] Retrying with prompt nudge...");
                        session.Messages.Add(new ChatRequestUserMessage("Please provide a complete text response summarizing your answer."));
                        continue;
                    }
                    else
                    {
                        finalText = "(empty response)";
                    }
                }

                if (session.HasFileMutations && !session.HasExecutedVerification && session.PreVerifyNudges < 2)
                {
                    session.PreVerifyNudges++;
                    Console.WriteLine("  [Pre-Verify Quality Gate] Intercepted completed turn. Files modified without verification. Prompting agent to verify...");
                    session.Messages.Add(new ChatRequestAssistantMessage(finalText));
                    session.Messages.Add(new ChatRequestUserMessage("You modified files during this turn. Please verify your changes by executing unit tests or build commands before concluding."));
                    continue;
                }

                session.Messages.Add(new ChatRequestAssistantMessage(finalText));
                swTurn.Stop();
                AgentTelemetry.TrackTurnEnd(session.SessionId, session.UserId, session.TurnCount, swTurn.ElapsedMilliseconds, finalText, "completed", session.SessionId, turnSpanId);
                return new TurnResult { FinalResponse = finalText, Messages = session.Messages, ApiCalls = apiCalls, Completed = true, ExitReason = "text_response" };
            }

            Console.WriteLine($"  [Turn Exit] Reached max iterations ({config.MaxIterations}).");
            swTurn.Stop();
            var maxIterRes = new TurnResult { FinalResponse = "Reached maximum iteration limit.", Messages = session.Messages, ApiCalls = apiCalls, Failed = true, ExitReason = "budget_exhausted" };
            AgentTelemetry.TrackTurnEnd(session.SessionId, session.UserId, session.TurnCount, swTurn.ElapsedMilliseconds, maxIterRes.FinalResponse, maxIterRes.ExitReason, session.SessionId, turnSpanId);
            return maxIterRes;
        }

        private void DrainSteering(AgentSessionState session)
        {
            if (session.SteeringQueue.IsEmpty) return;

            var items = new List<string>();
            while (session.SteeringQueue.TryDequeue(out var item))
            {
                if (!string.IsNullOrWhiteSpace(item))
                {
                    items.Add(item);
                }
            }

            if (items.Count == 0) return;

            string steeringContent = "\n\n[USER STEERING INTERRUPT]: " + string.Join("\n", items);
            Console.WriteLine($"  [Pre-API Steering Drain] Injected {items.Count} mid-turn steering message(s)");

            if (session.Messages.Count > 0)
            {
                var lastMsg = session.Messages[session.Messages.Count - 1];
                if (lastMsg is ChatRequestToolMessage toolMsg)
                {
                    session.Messages[session.Messages.Count - 1] = new ChatRequestToolMessage(toolMsg.Content + steeringContent, toolMsg.ToolCallId);
                }
                else if (lastMsg is ChatRequestUserMessage userMsg)
                {
                    session.Messages[session.Messages.Count - 1] = new ChatRequestUserMessage(userMsg.Content + steeringContent);
                }
                else
                {
                    session.Messages.Add(new ChatRequestUserMessage("[USER STEERING INTERRUPT]: " + string.Join("\n", items)));
                }
            }
            else
            {
                session.Messages.Add(new ChatRequestUserMessage("[USER STEERING INTERRUPT]: " + string.Join("\n", items)));
            }
        }

        public static string SanitizeJsonArguments(string toolName, string rawArgs)
        {
            if (string.IsNullOrWhiteSpace(rawArgs)) return "{}";
            string clean = rawArgs.Trim();

            // 1. Extract JSON object if concatenated with leading tokens (e.g. "types.fs" {"path":"types.fs"})
            int braceIndex = clean.IndexOf('{');
            if (braceIndex > 0)
            {
                int lastBrace = clean.LastIndexOf('}');
                if (lastBrace > braceIndex)
                {
                    clean = clean.Substring(braceIndex, lastBrace - braceIndex + 1);
                }
            }

            // 2. Unescape double-encoded JSON strings
            if (clean.StartsWith("\"") && clean.EndsWith("\""))
            {
                try
                {
                    var unescaped = JsonSerializer.Deserialize<string>(clean);
                    if (!string.IsNullOrWhiteSpace(unescaped)) clean = unescaped.Trim();
                }
                catch { }
            }

            // 3. Re-check brace extraction after unescaping
            braceIndex = clean.IndexOf('{');
            if (braceIndex >= 0)
            {
                int lastBrace = clean.LastIndexOf('}');
                if (lastBrace > braceIndex)
                {
                    clean = clean.Substring(braceIndex, lastBrace - braceIndex + 1);
                }
            }

            // 4. Handle concatenated JSON objects e.g. {"path":"Telemetry.fs"}{"command":"ls"}
            try
            {
                using var testDoc = JsonDocument.Parse(clean);
            }
            catch
            {
                int lastFirstBrace = clean.LastIndexOf('{');
                int lastLastBrace = clean.LastIndexOf('}');
                if (lastFirstBrace > 0 && lastLastBrace > lastFirstBrace)
                {
                    var candidate = clean.Substring(lastFirstBrace, lastLastBrace - lastFirstBrace + 1);
                    try
                    {
                        using var candidateDoc = JsonDocument.Parse(candidate);
                        clean = candidate;
                    }
                    catch { }
                }
            }

            // 4. If clean is a raw unquoted string or file path without braces (e.g. "types.fs")
            if (!clean.StartsWith("{"))
            {
                var rawString = clean.Trim('"');
                clean = JsonSerializer.Serialize(new Dictionary<string, string>
                {
                    ["path"] = rawString,
                    ["command"] = rawString,
                    ["key"] = rawString,
                    ["task"] = rawString,
                    ["text"] = rawString,
                    ["url"] = rawString
                });
            }

            return clean;
        }
    }
}
